akka-docs/rst/java/code/docs/stream/io/StreamTcpDocTest.java

```diff
@@ -95,50 +98,33 @@ public void actuallyWorkingClientServerApp() {
     final TestProbe serverProbe = new TestProbe(system);
 
     final Source<IncomingConnection,CompletionStage<ServerBinding>> connections =
-        Tcp.get(system).bind(localhost.getHostName(), localhost.getPort()); // TODO getHostString in Java7
+        Tcp.get(system).bind(localhost.getHostString(), localhost.getPort());
     //#welcome-banner-chat-server
     connections.runForeach(connection -> {
       // server logic, parses incoming commands
-      final PushStage<String, String> commandParser = new PushStage<String, String>() {
-        @Override public SyncDirective onPush(String elem, Context<String> ctx) {
-          if (elem.equals("BYE"))
-            return ctx.finish();
-          else
-            return ctx.push(elem + "!");
-        }
-      };
+      final Flow<String, String, NotUsed> commandParser =
+          Flow.<String>create()
+            .takeWhile(elem -> !elem.equals("BYE"))
+            .map(elem -> elem + "!");
 
       final String welcomeMsg = "Welcome to: " + connection.localAddress() +
-          " you are: " + connection.remoteAddress() + "!\n";
+          " you are: " + connection.remoteAddress() + "!";
 
-      final Source<ByteString, NotUsed> welcome =
-          Source.single(ByteString.fromString(welcomeMsg));
-      final Flow<ByteString, ByteString, NotUsed> echoFlow =
+      final Source<String, NotUsed> welcome = Source.single(welcomeMsg);
+      final Flow<ByteString, ByteString, NotUsed> serverLogic =
           Flow.of(ByteString.class)
             .via(Framing.delimiter(ByteString.fromString("\n"), 256, false))
-            .map(bytes -> bytes.utf8String())
+            .map(ByteString::utf8String)
             //#welcome-banner-chat-server
             .map(command -> {
               serverProbe.ref().tell(command, null);
               return command;
             })
             //#welcome-banner-chat-server
-            .transform(() -> commandParser)
+            .via(commandParser)
+            .merge(welcome)
             .map(s ->  s + "\n")
-            .map(s -> ByteString.fromString(s));
-
-      final Flow<ByteString, ByteString, NotUsed> serverLogic =
```

Comment:
Getting away with not using `GraphDSL` in this sample was a bit out of scope, but did it anyways because I saw how much simpler it could be. The docs where this is used are about IO and not the `GraphDSL`.